core/src/main/scala/com/criteo/cuttle/Workflow.scala

```diff
@@ -16,30 +17,144 @@ trait Workflow[S <: Scheduling] {
   private[criteo] def vertices: Set[Job[S]]
   private[criteo] def edges: Set[Dependency]
 
-  // Kahn's algorithm
-  // We construct the best linear representation for our DAG. At the
-  // same time it checks that there is no cycle.
-  private[cuttle] lazy val jobsInOrder: List[Job[S]] = {
-    val result = collection.mutable.ListBuffer.empty[Job[S]]
-    val edges = collection.mutable.Set(this.edges.toSeq: _*)
-    val orderedRoots = collection.mutable.SortedSet(roots.toSeq: _*)(Ordering.by(_.id))
-    while (orderedRoots.nonEmpty) {
-      val root = orderedRoots.head
-      orderedRoots.remove(root)
-      result.append(root)
-      val edgesToRemove = edges.filter(_._2 == root)
-
-      edgesToRemove.foreach {
-        case edge @ (child, _, _) =>
-          edges.remove(edge)
-          if (!edges.exists(_._1 == child)) {
-            orderedRoots.add(child)
-          }
+  def roots: Set[Job[S]] = {
+    val childNodes = edges.map { case (child, _, _) => child }
+    vertices.filter(!childNodes.contains(_))
+  }
+
+  def leaves: Set[Job[S]] = {
+    val parentNodes = edges.map { case (_, parent, _) => parent }
+    vertices.filter(!parentNodes.contains(_))
+  }
+
+  // Returns a list of jobs in the workflow sorted topologically, using Kahn's algorithm. At the
+  // same time checks that there is no cycle.
+  private[cuttle] lazy val jobsInOrder: List[Job[S]] = topologicalSort((_, _) => ()) match {
+    case Some(sortedJobs) => sortedJobs
+    case None => throw new IllegalArgumentException("Workflow has at least one cycle")
+  }
+
+  /**
+    * @return the list of strongly connected components (SCCs) in the workflow with more than 1 node, which
+    * are the cycles in the graph
+    * @note Tarjan algorithm
+    */
+  def findCycles(): List[List[Job[S]]] = {
```

Comment:
could be nice to implement it as a generic & pure function and call it here.
Not sure it we want to expose this in the API...

Reply:
Agreed, will do